openmc/mesh.py

```diff
@@ -1607,83 +1654,107 @@ def write_data_to_vtk(self, filename, datasets, volume_normalization=True):
+            elif elem_type == self._LINEAR_HEX:
+                elem = vtk.vtkHexahedron()
+            elif elem_type == self._UNSUPPORTED_ELEM:
+                n_skipped += 1
+            else:
+                raise RuntimeError(f'Invalid element type {elem_type} found')
+            for i, c in enumerate(conn):
+                if c == -1:
+                    break
+                elem.GetPointIds().SetId(i, c)
+            elems.append(elem)
 
-        # create a VTK data object
-        poly_data = vtk.vtkPolyData()
-        poly_data.SetPoints(points)
-        poly_data.SetVerts(vertices)
+        if n_skipped > 0:
+            warnings.warn(f'{n_skipped} elements were not written because '
+                          'they are not of type linear tet/hex')
 
-        # strange VTK nuance:
-        # data must be held in some container
-        # until the vtk file is written
-        data_holder = []
+        for elem in elems:
+            grid.InsertNextCell(elem.GetCellType(), elem.GetPointIds())
 
-        # create VTK arrays for each of
-        # the data sets
-        for label, dataset in datasets.items():
-            dataset = np.asarray(dataset).flatten()
+        # check that datasets are the correct size
+        datasets_out = []
+        if datasets is not None:
+            for name, data in datasets.items():
+                if data.shape != self.dimension:
+                    raise ValueError(f'Cannot apply dataset "{name}" with '
+                                     f'shape {data.shape} to mesh {self.id} '
+                                     f'with dimensions {self.dimension}')
 
             if volume_normalization:
-                dataset /= self.volumes.flatten()
-
-            array = vtk.vtkDoubleArray()
-            array.SetName(label)
-            array.SetNumberOfComponents(1)
-            array.SetArray(vtk_npsup.numpy_to_vtk(dataset),
-                           dataset.size,
-                           True)
-
-            data_holder.append(dataset)
-            poly_data.GetPointData().AddArray(array)
-
-        # set filename
-        if not filename.endswith(".vtk"):
-            filename += ".vtk"
+                for name, data in datasets.items():
+                    if np.issubdtype(data.dtype, np.integer):
+                        warnings.warn(f'Integer data set "{name}" will '
+                                      'not be volume-normalized.')
+                        continue
+                    data /= self.volumes
+
+            # add data to the mesh
+            for name, data in datasets.items():
+                datasets_out.append(data)
+                arr = vtk.vtkDoubleArray()
+                arr.SetName(name)
+                arr.SetNumberOfTuples(data.size)
+
+                for i in range(data.size):
+                    arr.SetTuple1(i, data.flat[i])
+                grid.GetCellData().AddArray(arr)
+
+        if vtk.VTK_MAJOR_VERSION == 5:
```

Comment:
Would `>=` make sense?

Reply:
Good point. VTK is on [version 9.2.0 now](https://pypi.org/project/vtk/#history), so I think it should be `<=`.

Reply:
Alternatively, we might not worry about this at all since versions `<= 5` are so old now?